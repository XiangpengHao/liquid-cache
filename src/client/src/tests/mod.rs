use std::{path::Path, sync::Arc};

use datafusion::{
    physical_plan::{ExecutionPlan, display::DisplayableExecutionPlan},
    prelude::{SessionConfig, SessionContext},
};

use crate::LiquidCacheBuilder;

async fn setup_tpch_ctx() -> Arc<SessionContext> {
    let mut session_config = SessionConfig::from_env().unwrap();
    // To get deterministic results
    session_config.options_mut().execution.target_partitions = 8;
    let ctx = LiquidCacheBuilder::new("http://localhost:50051")
        .build(session_config)
        .unwrap();
    let tables = [
        "customer", "lineitem", "nation", "orders", "part", "partsupp", "region", "supplier",
    ];

    for table in tables {
        let table_url = format!("../../benchmark/tpch/data/sf0.001/{}.parquet", table);
        ctx.register_parquet(table, table_url, Default::default())
            .await
            .unwrap();
    }
    Arc::new(ctx)
}

/// One query file can contain multiple queries, separated by `;`
fn get_query_by_id(query_dir: impl AsRef<Path>, query_id: u32) -> Vec<String> {
    let query_dir = query_dir.as_ref();
    let mut path = query_dir.to_owned();
    path.push(format!("q{query_id}.sql"));
    let content = std::fs::read_to_string(&path).unwrap();
    content
        .split(';')
        .map(|s| s.trim())
        .filter(|s| !s.is_empty())
        .map(|s| s.to_string())
        .collect()
}

async fn get_query_plan_by_id(query_id: u32) -> Arc<dyn ExecutionPlan> {
    let query_dir = Path::new("../../benchmark/tpch/queries");
    let ctx = setup_tpch_ctx().await;

    let query = if query_id != 15 {
        &get_query_by_id(query_dir, query_id)[0]
    } else {
        &get_query_by_id(query_dir, query_id)[1]
    };
    let result = ctx.sql(query).await.unwrap();
    let (state, plan) = result.into_parts();
    state.create_physical_plan(&plan).await.unwrap()
}

#[tokio::test]
async fn test_tpch_q1() {
    let physical_plan = get_query_plan_by_id(1).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q2() {
    let physical_plan = get_query_plan_by_id(2).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q3() {
    let physical_plan = get_query_plan_by_id(3).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q4() {
    let physical_plan = get_query_plan_by_id(4).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q5() {
    let physical_plan = get_query_plan_by_id(5).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q6() {
    let physical_plan = get_query_plan_by_id(6).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q7() {
    let physical_plan = get_query_plan_by_id(7).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q8() {
    let physical_plan = get_query_plan_by_id(8).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q9() {
    let physical_plan = get_query_plan_by_id(9).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q10() {
    let physical_plan = get_query_plan_by_id(10).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q11() {
    let physical_plan = get_query_plan_by_id(11).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q12() {
    let physical_plan = get_query_plan_by_id(12).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q13() {
    let physical_plan = get_query_plan_by_id(13).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q14() {
    let physical_plan = get_query_plan_by_id(14).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
/// Q15 needs to actually run the first query to get the correct plan
#[ignore]
async fn test_tpch_q15() {
    let physical_plan = get_query_plan_by_id(15).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q16() {
    let physical_plan = get_query_plan_by_id(16).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q17() {
    let physical_plan = get_query_plan_by_id(17).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q18() {
    let physical_plan = get_query_plan_by_id(18).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q19() {
    let physical_plan = get_query_plan_by_id(19).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q20() {
    let physical_plan = get_query_plan_by_id(20).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q21() {
    let physical_plan = get_query_plan_by_id(21).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q22() {
    let physical_plan = get_query_plan_by_id(22).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}
