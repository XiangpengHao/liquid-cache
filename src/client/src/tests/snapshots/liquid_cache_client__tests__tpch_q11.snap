---
source: src/client/src/tests/mod.rs
expression: displayable.tree_render().to_string()
---
┌───────────────────────────┐
│  SortPreservingMergeExec  │
│    --------------------   │
│         value DESC        │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│          SortExec         │
│    --------------------   │
│        value@1 DESC       │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│       ProjectionExec      │
│    --------------------   │
│        ps_partkey:        │
│         ps_partkey        │
│                           │
│           value:          │
│ sum(partsupp.ps_supplycost│
│   * partsupp.ps_availqty) │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│     NestedLoopJoinExec    ├────────────────────────────────────────────────────────────────────────┐
└─────────────┬─────────────┘                                                                        │
┌─────────────┴─────────────┐                                                          ┌─────────────┴─────────────┐
│       ProjectionExec      │                                                          │       AggregateExec       │
│    --------------------   │                                                          │    --------------------   │
│ sum(partsupp.ps_supplycost│                                                          │           aggr:           │
│  * partsupp.ps_availqty)  │                                                          │ sum(partsupp.ps_supplycost│
│     * Float64(0.0001):    │                                                          │   * partsupp.ps_availqty) │
│   CAST(CAST(sum(partsupp  │                                                          │                           │
│      .ps_supplycost *     │                                                          │    group_by: ps_partkey   │
│          partsupp         │                                                          │                           │
│       .ps_availqty)       │                                                          │           mode:           │
│   AS Float64) * 0.0001 AS │                                                          │      FinalPartitioned     │
│     Decimal128(38, 15))   │                                                          │                           │
└─────────────┬─────────────┘                                                          └─────────────┬─────────────┘
┌─────────────┴─────────────┐                                                          ┌─────────────┴─────────────┐
│       AggregateExec       │                                                          │    CoalesceBatchesExec    │
│    --------------------   │                                                          │    --------------------   │
│           aggr:           │                                                          │     target_batch_size:    │
│ sum(partsupp.ps_supplycost│                                                          │           16384           │
│   * partsupp.ps_availqty) │                                                          │                           │
│                           │                                                          │                           │
│        mode: Final        │                                                          │                           │
└─────────────┬─────────────┘                                                          └─────────────┬─────────────┘
┌─────────────┴─────────────┐                                                          ┌─────────────┴─────────────┐
│   CoalescePartitionsExec  │                                                          │      RepartitionExec      │
│                           │                                                          │    --------------------   │
│                           │                                                          │ partition_count(in->out): │
│                           │                                                          │           8 -> 8          │
│                           │                                                          │                           │
│                           │                                                          │    partitioning_scheme:   │
│                           │                                                          │  Hash([ps_partkey@0], 8)  │
└─────────────┬─────────────┘                                                          └─────────────┬─────────────┘
┌─────────────┴─────────────┐                                                          ┌─────────────┴─────────────┐
│       AggregateExec       │                                                          │       AggregateExec       │
│    --------------------   │                                                          │    --------------------   │
│           aggr:           │                                                          │           aggr:           │
│ sum(partsupp.ps_supplycost│                                                          │ sum(partsupp.ps_supplycost│
│   * partsupp.ps_availqty) │                                                          │   * partsupp.ps_availqty) │
│                           │                                                          │                           │
│       mode: Partial       │                                                          │    group_by: ps_partkey   │
│                           │                                                          │       mode: Partial       │
└─────────────┬─────────────┘                                                          └─────────────┬─────────────┘
┌─────────────┴─────────────┐                                                          ┌─────────────┴─────────────┐
│    CoalesceBatchesExec    │                                                          │    CoalesceBatchesExec    │
│    --------------------   │                                                          │    --------------------   │
│     target_batch_size:    │                                                          │     target_batch_size:    │
│           16384           │                                                          │           16384           │
└─────────────┬─────────────┘                                                          └─────────────┬─────────────┘
┌─────────────┴─────────────┐                                                          ┌─────────────┴─────────────┐
│        HashJoinExec       │                                                          │        HashJoinExec       │
│    --------------------   │                                                          │    --------------------   │
│            on:            ├──────────────┐                                           │            on:            ├──────────────┐
│(n_nationkey = s_nationkey)│              │                                           │(n_nationkey = s_nationkey)│              │
└─────────────┬─────────────┘              │                                           └─────────────┬─────────────┘              │
┌─────────────┴─────────────┐┌─────────────┴─────────────┐                             ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
│   LiquidCacheClientExec   ││      RepartitionExec      │                             │   LiquidCacheClientExec   ││       ProjectionExec      │
│    --------------------   ││    --------------------   │                             │    --------------------   ││    --------------------   │
│          server:          ││ partition_count(in->out): │                             │          server:          ││        ps_availqty:       │
│  http://localhost:50051,  ││           1 -> 8          │                             │  http://localhost:50051,  ││        ps_availqty        │
│        mode=liquid,       ││                           │                             │        mode=liquid,       ││                           │
│      object_stores=[]     ││    partitioning_scheme:   │                             │      object_stores=[]     ││        ps_partkey:        │
│                           ││     RoundRobinBatch(8)    │                             │                           ││         ps_partkey        │
│                           ││                           │                             │                           ││                           │
│                           ││                           │                             │                           ││       ps_supplycost:      │
│                           ││                           │                             │                           ││       ps_supplycost       │
│                           ││                           │                             │                           ││                           │
│                           ││                           │                             │                           ││        s_nationkey:       │
│                           ││                           │                             │                           ││        s_nationkey        │
└─────────────┬─────────────┘└─────────────┬─────────────┘                             └─────────────┬─────────────┘└─────────────┬─────────────┘
┌─────────────┴─────────────┐┌─────────────┴─────────────┐                             ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
│       DataSourceExec      ││       ProjectionExec      │                             │       DataSourceExec      ││    CoalesceBatchesExec    │
│    --------------------   ││    --------------------   │                             │    --------------------   ││    --------------------   │
│          files: 1         ││        ps_availqty:       │                             │          files: 1         ││     target_batch_size:    │
│      format: parquet      ││        ps_availqty        │                             │      format: parquet      ││           16384           │
│                           ││                           │                             │                           ││                           │
│         predicate:        ││       ps_supplycost:      │                             │         predicate:        ││                           │
│      n_name = GERMANY     ││       ps_supplycost       │                             │      n_name = GERMANY     ││                           │
│                           ││                           │                             │                           ││                           │
│                           ││        s_nationkey:       │                             │                           ││                           │
│                           ││        s_nationkey        │                             │                           ││                           │
└───────────────────────────┘└─────────────┬─────────────┘                             └───────────────────────────┘└─────────────┬─────────────┘
                             ┌─────────────┴─────────────┐                                                          ┌─────────────┴─────────────┐
                             │    CoalesceBatchesExec    │                                                          │        HashJoinExec       │
                             │    --------------------   │                                                          │    --------------------   │
                             │     target_batch_size:    │                                                          │            on:            ├──────────────┐
                             │           16384           │                                                          │  (s_suppkey = ps_suppkey) │              │
                             └─────────────┬─────────────┘                                                          └─────────────┬─────────────┘              │
                             ┌─────────────┴─────────────┐                                                          ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
                             │        HashJoinExec       │                                                          │   LiquidCacheClientExec   ││   LiquidCacheClientExec   │
                             │    --------------------   │                                                          │    --------------------   ││    --------------------   │
                             │            on:            │                                                          │          server:          ││          server:          │
                             │  (s_suppkey = ps_suppkey) ├──────────────┐                                           │  http://localhost:50051,  ││  http://localhost:50051,  │
                             │                           │              │                                           │        mode=liquid,       ││        mode=liquid,       │
                             │                           │              │                                           │      object_stores=[]     ││      object_stores=[]     │
                             └─────────────┬─────────────┘              │                                           └─────────────┬─────────────┘└─────────────┬─────────────┘
                             ┌─────────────┴─────────────┐┌─────────────┴─────────────┐                             ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
                             │   LiquidCacheClientExec   ││   LiquidCacheClientExec   │                             │       DataSourceExec      ││      RepartitionExec      │
                             │    --------------------   ││    --------------------   │                             │    --------------------   ││    --------------------   │
                             │          server:          ││          server:          │                             │          files: 1         ││ partition_count(in->out): │
                             │  http://localhost:50051,  ││  http://localhost:50051,  │                             │      format: parquet      ││           1 -> 8          │
                             │        mode=liquid,       ││        mode=liquid,       │                             │                           ││                           │
                             │      object_stores=[]     ││      object_stores=[]     │                             │                           ││    partitioning_scheme:   │
                             │                           ││                           │                             │                           ││     RoundRobinBatch(8)    │
                             └─────────────┬─────────────┘└─────────────┬─────────────┘                             └───────────────────────────┘└─────────────┬─────────────┘
                             ┌─────────────┴─────────────┐┌─────────────┴─────────────┐                                                          ┌─────────────┴─────────────┐
                             │       DataSourceExec      ││       DataSourceExec      │                                                          │       DataSourceExec      │
                             │    --------------------   ││    --------------------   │                                                          │    --------------------   │
                             │          files: 1         ││          files: 1         │                                                          │          files: 1         │
                             │      format: parquet      ││      format: parquet      │                                                          │      format: parquet      │
                             │                           ││                           │                                                          │      predicate: true      │
                             └───────────────────────────┘└───────────────────────────┘                                                          └───────────────────────────┘
